### DISCUSSION TITLE
RLS Should Log Rejections

### DISCUSSION THREAD

**Discussion by caseycrogers** [supabase/supabase#9214] [Feature Requests]
Right now, there are situations where a request will be silently rejected by RLS. The client will receive a 200 with empty content and the server logs will report the same when what happened was the action was not executed because it violated an RLS rule. I understand the need to return a misleading 200 to prevent information leaks to malicious actors, but the lack of server logs makes it really hard to debug failing requests.

The server logs should log an error message when RLS rejects a request so that developers can quickly root cause a request that appears to succeed but produces no results.

See related issue:
https://github.com/supabase-community/supabase-flutter/issues/175

**Comment by steve-chavez** [comment]
> The client will receive a 200 with empty content and the server logs will report the same

If you expect a client to get a single row you could use [.single()](https://supabase.com/docs/reference/javascript/single) and this will fail at the client side plus logged server side with a 406 status(Not Acceptable).

Would that solve your problem?

**Comment by caseycrogers** [comment]
Nah, as at least in the specific circumstances I've run into this I've been running delete statements and wasted an hour or so trying to figure out why my delete was ostensibly successful (200, no errors int he server side logs) but the record wasn't getting deleted.

**Reply by steve-chavez** [reply]
Note that `delete()` works with `single()` too, the same server-side 406 logging would apply as well.

**Reply by steve-chavez** [reply]
Instead of returning `200` with an empty array body `[]`, would returning a `204 No Content` with no body help?

**Reply by caseycrogers** [reply]
204 no content would still suggest that the request had completed, so I would still fall down the same confusing rabbit hole debugging why an ostensibly successful request was having no effect. The `single()` and `406` strategy might work, but the issue is this is awkward and non-intuitive: every new Supabase user is going to hit this bug, waste a lot of time on it, and then find out (and be frustrated) that RLS returns 200's on rejected requests.

IMO ideal behavior would be as follows:
1) RLS-rejected requests would return non-200 responses
1a) RLS should allow developers to set custom failure messages server-side for the response body
      ^ When I worked at Google I built this for Firebase RTDB security rules as a part of a hackathon, it was such a cool feature in internal testing! Sadly it never saw the light of day ):
2) The server logs explicit and detailed messages that tell the developer exactly what rule failed, and what the relevant state was that failed. And/or a specific "rules" dashboard that shows me in an interactive fashion what rules were most recently tested, which ones failed or succeeded, and why.

My understanding (correct me if I'm wrong?) is that (1) isn't possible because it creates an information leak where a malicious actor may be able to tell if a record exists or not even if they don't have read access. (1a) may actually be a great work around for this-if the developer has specified a non-empty error message, then you can deliver that error message and the developer is culpable for any information leaks returning a non-200 may cause. For example, my entire DB is public read, only writes are restricted. For me, there is no potential for information leaks so I'd like as explicit of error messages as possible. This way I could opt-in to explicit and customized error messages.
(2) seems like bare-minimum behavior if (1) and/or (1a) can't be done.

**Reply by anngbaum** [reply]
> The server logs explicit and detailed messages that tell the developer exactly what rule failed, and what the relevant state was that failed. And/or a specific "rules" dashboard that shows me in an interactive fashion what rules were most recently tested, which ones failed or succeeded, and why.

just jumping in again because clearer logs would have been a HUGE help in figuring this out – even just at the level of connecting the `edge_logs` table (which has the API response code) to `postgres_logs` (which contained some of the authentication info) was really challenging! adding authentication status to the `edge_logs` would be a super easy win here. I ended up basically joining on timestamp to connect those dots (detailed further in discord [here](https://discord.com/channels/839993398554656828/989219603399974923/993955089800773824)), which was clunky.

**Reply by Ziinc** [reply]
@anngbaum  RE supabase logging, We're working on getting joins working between `postgres_logs` and `edge_logs`, you can follow along the progress on this PR https://github.com/PostgREST/postgrest/issues/2506

**Comment by anngbaum** [comment]
just a +1 here having spent a **lot** of time trying to debug whether a 406 in a `.maybeSingle()` request was mysteriously failing because no record was found or because there was an authentication issue due to RLS. would have been extremely helpful to get a 401 forbidden response to isolate authentication issues. full discussion was in discord [here](https://discord.com/channels/839993398554656828/989219603399974923/989231048275673170) for reference.

**Reply by steve-chavez** [reply]
> no record was found or because there was an authentication issue due to RLS

Internally I'm proposing to drop the `anon` role so if Auth(GoTrue) somehow fails to authenticate the request it won't go through to the db as the `anon` role and execute RLS - instead the request will be rejected at the REST level with a 401.

What do you think? Would that help?

**Reply by anngbaum** [reply]
that makes a lot of sense to me!

[editing this because I misread the first part of your suggestion but realize anon requests would still work fine]

generally I think whenever we **do attach a token** we would expect an auth failure to cause the entire request to fail with a `401` rather than an attempted `anon` query that fails unexpectedly.

**Comment by silkgh** [comment]
I just spent nearly 2 hours working out that RLS wasn't working because nothing appears in the logs to say the update was rejected. This really needs to be sorted out please.

**Reply by GaryAustin1** [reply]
Unless Postgres itself changes to generate something not sure how the Supabase API would know unless you are using .select() to return a result.  With .select() you would get updated or deleted data back.  

RLS is part of Postgres and is silent for reads (which impact update and deletes) the same as filters not being met.
